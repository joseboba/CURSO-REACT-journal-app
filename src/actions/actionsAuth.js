import { firebase, googleAuthProvider } from '../firebase/firebase-config'
import Swal from 'sweetalert2'
import { types } from "../types/types"
import { finishLoading, startLoading } from './actionsUi'
import { noteLogout } from './actionsNotes'


export const startLoginEmailPassword = (email, password) => {
    
    return (dispatch) => {
        
        dispatch(startLoading())
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({user}) => {
                dispatch(login(user.uid , user.displayName)) 
                dispatch(finishLoading())
            })
            .catch(e =>{
                console.log(e)
                dispatch(finishLoading())
                Swal.fire('Error', e.message, 'error')
            })
    }

}

export const startRegisterWithEmailPasswordName = (email, password, name) => {

  
    return( dispatch ) => {

        firebase.auth().createUserWithEmailAndPassword( email, password )
        .then( async({ user }) => {
           await user.updateProfile({ displayName: name })
            dispatch(
                login(user.uid, user.displayName)
            )
        })
        .catch(e => {
            Swal.fire('Error', e.message, 'error')
        })

    }
}

export const startGoogleLogin = () => {
    return ( dispatch ) => {
        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( ({ user }) => {
                const { uid, displayName } = user;
                dispatch( login(uid, displayName) )
            })
    }
}


export const login = (uid, displayName) => {

    return{
        type: types.login,
        payload: {
            uid, 
            displayName
        }
    }

}

export const startLogout = () => {
    return async(dispatch) => {
        await firebase.auth().signOut();

        dispatch( logout() )
        dispatch( noteLogout() );
    }
}

export const logout = () => {
    return {
        type: types.logout
    }
}