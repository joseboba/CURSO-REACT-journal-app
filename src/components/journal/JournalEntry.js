import React from 'react'
import moment from 'moment'
import 'moment/locale/es';
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/actionsNotes';
export const JournalEntry = ({ id, title, body, url, date }) => {
    
    const dispatch = useDispatch();
    
    const noteDate = moment(date);
    
    const note = {
        title,
        body, 
        url, 
        date
    }

    const handleEntryClick  = (e) => {
        e.preventDefault();
        dispatch(activeNote( id, note ))
    }

    return (
        <div  className="journal__entry pointer animate__animated animate__backInLeft" onClick={ handleEntryClick }>

        {    
            url && 

            <div className="journal__entry-picture" style={{
                    backgroundSize: 'cover',
                    backgroundImage: `url(${url})`
                }}>
            
            </div>
        }
            <div className="journal__entry-body" >
                <p className="journal__entry-title"> { title } </p>
                <p className="journal__entry-content"> { body } </p>
            </div>

            <div className="journal__entry-date-box">
                <span> { noteDate.format('MMMM, dddd ') } </span>
                <h4>{ noteDate.format('DD') }</h4>
            </div>
        </div>
    )
}
