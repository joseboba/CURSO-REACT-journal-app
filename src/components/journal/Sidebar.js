import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startLogout } from '../../actions/actionsAuth'
import { startNewNote } from '../../actions/actionsNotes'

import { JournalEntries } from './JournalEntries'

export const Sidebar = () => {

    
    const dispatch = useDispatch()
    const { auth }  = useSelector(state => state)
    
    const handeLogout = () => {
        dispatch(startLogout());
    }

    const handleAddNew = () => {
        dispatch(startNewNote());
    }

    return (
        <aside className="journal_side-bar">
            <div className="journal_side-navbar">
                <h3 className="mt-5">
                    <i className="far fa-moon"></i>
                    <span> {`${ auth.name }`} </span>
                </h3>

                <button className="btn" onClick={ handeLogout } >
                    Cerrar SesiÃ³n
                </button>
            </div>

            <div  className="journal_new-entry" onClick={ handleAddNew } > 
                <i className="far fa-calendar-plus fa-5x" ></i>
                <p className="mt-5">
                    Nueva entrada
                </p>
            </div>

            <JournalEntries />

        </aside>
    )
}
