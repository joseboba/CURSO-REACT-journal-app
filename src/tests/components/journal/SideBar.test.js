import '@testing-library/jest-dom';
import React from 'react';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store' //ES6 modules
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { Sidebar } from '../../../components/journal/Sidebar';
import { startLogout } from '../../../actions/actionsAuth';
import { types } from '../../../types/types';
import { startNewNote } from '../../../actions/actionsNotes';



const middlewares = [thunk]
const mockStore = configureStore(middlewares);
const initState = {
    auth: {
        name: 'Jose Enrique'
    },
    ui: {
        lodgin: false,
        msgError: null
    },
    notes: {
        active: {
            id: 'ABC'
        },
        notes: []
    }
};
let store = mockStore(initState)
// store.dispatch = jest.fn();

describe('Pruebas en el SideBar', () => {

    beforeEach(() => {
        store = mockStore(initState);    
    })

    const wrapper = mount(
        <Provider store={ store }>
            <Sidebar />
        </Provider>
    )
    
    test('debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('debe de llamar el startLogout', async() => {
        //debe de llamar la accion del logut
        wrapper.find('button').simulate('click');
        await store.dispatch(startLogout())
        const action = await store.getActions()
        
        expect(action[0]).toEqual({ type: types.logout })
        expect(action[1]).toEqual({ type: types.notesLogoutCleaning });
    })

    test('debe de llamar el startNewNote', async() => {
        //debe de llamar la accion del logut
        wrapper.find('.journal_new-entry').simulate('click');
        await store.dispatch(startNewNote())
        const action = await store.getActions()
        
        expect(action[0]).toEqual({ 
            type: types.notesActive,
            payload: {
                id: expect.any(String),
                title: expect.any(String),
                body: expect.any(String),
                date: expect.any(Number)
            } 
        })
       
    })
    

})
