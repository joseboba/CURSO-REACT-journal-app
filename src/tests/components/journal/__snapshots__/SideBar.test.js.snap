// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el SideBar debe de mostrarse correctamente 1`] = `
<aside
  className="journal_side-bar"
>
  <div
    className="journal_side-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="far fa-moon"
      />
      <span>
         
        Jose Enrique
         
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Cerrar Sesi√≥n
    </button>
  </div>
  <div
    className="journal_new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x"
    />
    <p
      className="mt-5"
    >
      Nueva entrada
    </p>
  </div>
  <div
    className="journal__entries animate__animated animate__backInLeft "
  />
</aside>
`;
